core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -535,6 +536,103 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
+    val topic = "test"
+    val partition = 0
+    // create brokers
+    val serverConfigs = TestUtils.createBrokerConfigs(4, zkConnect, false, enableControlledShutdownSafetyCheck = true)
+      .map(KafkaConfig.fromProps)
+    servers = serverConfigs.reverseMap(s => TestUtils.createServer(s))
+    TestUtils.waitUntilControllerElected(zkClient)
+
+    // create the topic with min ISR of 2, which should allow one broker to shut down but should block subsequent
+    // shutdowns.
+    val topicConfig = new Properties()
+    topicConfig.setProperty(TopicConfig.MIN_IN_SYNC_REPLICAS_CONFIG, "2")
+    TestUtils.createTopic(zkClient, topic, partitionReplicaAssignment = expectedReplicaAssignment, servers = servers, topicConfig = topicConfig)
+
+    val controllerId = zkClient.getControllerId.get
+    val controller = servers.find(p => p.config.brokerId == controllerId).get.kafkaController
+    val resultQueue = new LinkedBlockingQueue[Try[collection.Set[TopicPartition]]]()
+
+    controller.setMinInSyncReplicas(topic, 2)
+    TestUtils.waitUntilTrue(() => controller.controllerContext.topicMinIsrConfig.getOrElse(topic, -1) == 2, "Controller never saw min.insync.replicas config change for topic.")
+
+    // Attempt to shut down one broker, which should be allowed with the ISR at 3 and the min ISR of 2
+    val controlledShutdownCallback = (controlledShutdownResult: Try[collection.Set[TopicPartition]]) => resultQueue.put(controlledShutdownResult)
+    controller.controlledShutdown(2, servers.find(_.config.brokerId == 2).get.kafkaController.brokerEpoch, controlledShutdownCallback)
+    var partitionsRemaining = resultQueue.take().get
+    var activeServers = servers.filter(s => s.config.brokerId != 2)
+    // wait for the update metadata request to trickle to the brokers
+    TestUtils.waitUntilTrue(() =>
+      activeServers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.isr.size != 4),
+      "Topic test not created after timeout")
+    assertEquals(0, partitionsRemaining.size)
+    var partitionStateInfo = activeServers.head.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get
+    var leaderAfterShutdown = partitionStateInfo.leader
+    assertEquals(0, leaderAfterShutdown)
+    assertEquals(3, partitionStateInfo.isr.size)
+    assertEquals(List(0, 1, 3), partitionStateInfo.isr.asScala)
+
+    // Now attempt to shut down a second broker which should fail with NotEnoughReplicasException
+    @volatile var notEnoughReplicasDetected = false
+    controller.controlledShutdown(1, servers.find(_.config.brokerId == 1).get.kafkaController.brokerEpoch, {
+      case scala.util.Failure(exception) if exception.isInstanceOf[NotEnoughReplicasException] => notEnoughReplicasDetected = true
+      case _ =>
+    })
+
+    TestUtils.waitUntilTrue(() => notEnoughReplicasDetected, "Fail to detect expected NotEnoughReplicasException")
+
+    val expectedShutdownEntries = Map(2 -> 49)
```

Comment:
Can you please explain why it'd be epoch 49?

Reply:
That's the epoch every time. I think it's just because this is a fresh zookeeper with a fixed number of clients that are connecting in the same order every time, we end up with the same brokerEpoch (which comes from `kafka.zk.KafkaZkClient#registerBroker`).

If I run this test with logging turned up you can see the brokers register in the same order every time, and that is the czxid from each registration.

```
[2021-01-26 21:37:37,580] INFO Registered broker 3 at path /brokers/ids/3 with addresses: ArrayBuffer(EndPoint(localhost,44075,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 27 (kafka.zk.KafkaZkClient:74)
[2021-01-26 21:37:37,874] INFO Registered broker 2 at path /brokers/ids/2 with addresses: ArrayBuffer(EndPoint(localhost,45451,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 49 (kafka.zk.KafkaZkClient:74)
[2021-01-26 21:37:37,966] INFO Registered broker 1 at path /brokers/ids/1 with addresses: ArrayBuffer(EndPoint(localhost,33383,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 69 (kafka.zk.KafkaZkClient:74)
[2021-01-26 21:37:38,040] INFO Registered broker 0 at path /brokers/ids/0 with addresses: ArrayBuffer(EndPoint(localhost,38473,ListenerName(PLAINTEXT),PLAINTEXT)), czxid (broker epoch): 89 (kafka.zk.KafkaZkClient:74)
```

Beyond that explaining why the epoch is always that is a zookeeper implementation detail.

Reply:
I see and thanks for the explanation. Then can you please put some comment here or add a link to this comment above?

Reply:
Done.